### ISSUE TITLE
Per-class metrics for multi-class classification and weighted macro-metrics

### ISSUE THREAD

**Issue by jklaise** [creme-ml/creme#145]
I'm wondering what's the canonical way of using `creme` to keep track of per-class metrics (e.g. accuracy, precision...) in multi-class classification?

Also, is there a plan to support metrics weighted by frequency of class labels? (e.g. `weighted` option in `sklearn`: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html#sklearn.metrics.f1_score).

**Comment by MaxHalford** [comment]
Hey there,

If you're well-versed in Python this is quite easy to do:

```python
import collections
from creme import metrics

class_metrics = collections.defaultdict(metrics.Precision)

for x, y in X_y:
    y_pred = model.predict_one(x)
    metrics[y].update(y, y_pred)

for label, metric in class_metrics.items():
    print(label, metric)
```

Then again not everyone is well-versed in Python, so maybe that this is something we should include in the API.

As for the weighted metrics, @gbolmier implemented the weighted version of F1-score in `metrics.MultiFBeta`, which supports a `weights` parameter. Are there any other metrics you would like to see with a weighted version?

**Comment by jklaise** [comment]
I thought about using a binary metric per class, but it's nice to see it written out, thanks!

As for the weighted metrics, what I had more in mind would be the facility by creme to keep track of online weights of the classes seen so far. Maybe one could achieve this with the existing API by keeping a track of individual per-class observation frequencies and then using those as weights for each update of the relevant metric?

**Comment by MaxHalford** [comment]
Indeed, this shouldn't be too hard to implement. I'll open a specific issue.

**Comment by MaxHalford** [comment]
Actually I was a bit hasty; we can keep the discussion here.

It would nice if some the metrics would support being weighted. That is, they should be able to take into account the support of each class (the number of times it appears). Effectively, this would be similar to macro averaging, except that the average should be weighted. I'm not exactly sure, but I think we can implement a single class called `metrics.Weighted`. The idea is that it would be passed a binary metric, and would take care of counting the label occurrence and doing the weighted average.

**Comment by jklaise** [comment]
That sounds good! I wonder if there are other such "modifiers" that one might want to apply to several metrics, I'm cautious that if there are, then it would be desirable to have composability (i.e. applying several at once) which might be tricky.

**Comment by MaxHalford** [comment]
Well, there is, namely "rolling" metrics. I remember having the same discussion with @AdilZouitine a few months ago, and we couldn't find any way to build a `Rolling` wrapper which worked. It would be really nice though, if we could do something like `Weighted(Rolling(Precision))`! It's definitely something very cool and useful to explore.

**Comment by jklaise** [comment]
Ah, of course. I can see how this would be difficult to design well, needs to be thought about more carefully.

**Comment by MaxHalford** [comment]
Hey @jklaise. I actually managed to implement a `Rolling` wrapper! It works really nicely too. I also added a `PerClass` wrapper for computing per-class metrics. Finally, every metric can now be passed a `sample_weight` parameter. All these changes are for the while in the [metrics-improvements](https://github.com/creme-ml/creme/tree/metrics-improvements) branch.

However, I wasn't able to implement a `Weighted` wrapper because the behavior seems to be different for precision and recall. I have to think about it!

**Comment by MaxHalford** [comment]
I just added `WeightedPrecision`, `WeightedRecall`, `WeightedFBeta`, and `WeightedF1`. I don't think it's possible to make a generic wrapper, but I would love to be proved wrong! I'll close this issue for now, but feel free to reopen it or to comment.
